crates/evm/src/execute.rs

```diff
@@ -0,0 +1,165 @@
+//! Traits for execution.
+
+use reth_interfaces::provider::ProviderError;
+use reth_primitives::U256;
+use revm::db::BundleState;
+use revm_primitives::db::Database;
+
+/// A general purpose executor trait that executes on an input (e.g. blocks) and produces an output
+/// (e.g. state changes and receipts).
+pub trait Executor<DB> {
+    /// The input type for the executor.
+    type Input<'a>;
+    /// The output type for the executor.
+    type Output;
+    /// The error type returned by the executor.
+    type Error;
+
+    /// Consumes the type and executes the block.
+    ///
+    /// Returns the output of the block execution.
+    fn execute(self, input: Self::Input<'_>) -> Result<Self::Output, Self::Error>;
+}
+
+/// An executor that can execute multiple blocks in a row and keep track of the state over the
+/// entire batch.
+pub trait BatchExecutor<DB> {
+    /// The input type for the executor.
+    type Input<'a>;
+    /// The output type for the executor.
+    type Output;
+    /// The error type returned by the executor.
+    type Error;
+
+    /// Executes the next block in the batch and update the state internally.
+    fn execute_one(&mut self, input: Self::Input<'_>) -> Result<BatchBlockOutput, Self::Error>;
+
+    /// Finishes the batch and return the final state.
+    fn finalize(self) -> Self::Output;
+}
+
+/// The output of an executed block in a batch.
+#[derive(Debug, Clone, Copy)]
+pub struct BatchBlockOutput {
+    /// The size hint of the batch's tracked state.
+    pub size_hint: Option<usize>,
+}
+
+/// The output of an ethereum block.
+///
+/// Contains the state changes, transaction receipts, and total gas used in the block.
+///
+/// TODO(mattsse): combine with BundleStateWithReceipts
+#[derive(Debug)]
+pub struct EthBlockOutput<T> {
+    /// The changed state of the block after execution.
+    pub state: BundleState,
+    /// All the receipts of the transactions in the block.
+    pub receipts: Vec<T>,
+    /// The total gas used by the block.
+    pub gas_used: u64,
```

Comment:
I'm good with this being concrete, would encourage us to proactively think about what will multi dimensional pricing look like in the future, e.g. thinking about "local fee markets" in parevm